tmva/tmva/src/RBDT.cxx

```diff
@@ -1,4 +1,376 @@
-#include "TMVA/RBDT.hxx"
+/**********************************************************************************
+ * Project: ROOT - a Root-integrated toolkit for multivariate data analysis       *
+ * Package: TMVA                                                                  *
+ *                                                                                *
+ *                                                                                *
+ * Description:                                                                   *
+ *                                                                                *
+ * Authors:                                                                       *
+ *      Jonas Rembser (jonas.rembser@cern.ch)                                     *
+ *                                                                                *
+ * Copyright (c) 2024:                                                            *
+ *      CERN, Switzerland                                                         *
+ *                                                                                *
+ * Redistribution and use in source and binary forms, with or without             *
+ * modification, are permitted according to the terms listed in LICENSE           *
+ * (see tmva/doc/LICENSE)                                          *
+ **********************************************************************************/
 
-template class TMVA::Experimental::RBDT<TMVA::Experimental::BranchlessForest<float>>;
-template class TMVA::Experimental::RBDT<TMVA::Experimental::BranchlessJittedForest<float>>;
+#include <TMVA/RBDT.hxx>
+
+#include <ROOT/RSpan.hxx>
+#include <ROOT/StringUtils.hxx>
+
+#include <TFile.h>
+#include <TSystem.h>
+
+#include <cmath>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <stdexcept>
+#include <stdlib.h>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace {
+
+namespace detail {
+
+typedef std::unordered_map<int, int> IndexMap;
+
+void correctIndices(std::span<int> indices, IndexMap const &nodeIndices, IndexMap const &leafIndices)
+{
+   for (int &idx : indices) {
+      auto foundNode = nodeIndices.find(idx);
+      if (foundNode != nodeIndices.end()) {
+         idx = foundNode->second;
+         continue;
+      }
+      auto foundLeaf = leafIndices.find(idx);
+      if (foundLeaf != leafIndices.end()) {
+         idx = -foundLeaf->second;
+         continue;
+      } else {
+         throw std::runtime_error("something is wrong in the node structure");
+      }
+   }
+}
+
+template <class Value_t>
+void softmaxTransformInplace(Value_t *out, int nOut)
+{
+   // Do softmax transformation inplace, mimicing exactly the Softmax function
+   // in the src/common/math.h source file of xgboost.
+   double norm = 0.;
+   Value_t wmax = *out;
+   int i = 1;
+   for (; i < nOut; ++i) {
+      wmax = std::max(out[i], wmax);
+   }
+   i = 0;
```

Comment:
Resetting the value of `i` inbetween for loops means that probably it should just be declared in the for loop directly